tests/Extensions/JobRetryExtensionTest.php

```diff
@@ -0,0 +1,179 @@
+<?php
+
+namespace SilverStripe\Forager\Tests\Extensions;
+
+use SilverStripe\Dev\SapphireTest;
+use SilverStripe\Forager\Tests\Fake\BatchJobFake;
+use SilverStripe\ORM\FieldType\DBDatetime;
+use Symbiote\QueuedJobs\DataObjects\QueuedJobDescriptor;
+use Symbiote\QueuedJobs\Services\QueuedJob;
+use Symbiote\QueuedJobs\Services\QueuedJobService;
+
+class JobRetryExtensionTest extends SapphireTest
+{
+
+    protected $usesDatabase = true; // phpcs:ignore SlevomatCodingStandard.TypeHints
+
+    /**
+     * We'll run the job a total of 4 times. We expect 3 attempts to break, but be retried, and we expect the 4th try
+     * to break and remain broken
+     *
+     * @dataProvider provideCodes
+     */
+    public function testRetryCodes(int $code): void
+    {
+        $job = new BatchJobFake($code);
+
+        $jobId = QueuedJobService::singleton()->queueJob($job);
+
+        // THE 1ST TRY
+        // This should break (but be allowed to retry)
+
+        // Run the job through our service, so that the extension point is invoked
+        QueuedJobService::singleton()->runJob($jobId);
+
+        // Fetch the Descriptor (after job processing) so that we can test that it was set with the correct data
+        $jobDescriptor = QueuedJobDescriptor::get()->byID($jobId);
+
+        $this->assertNotNull($jobDescriptor);
+
+        // The job status should have been set back to "New"
+        $this->assertEquals(QueuedJob::STATUS_NEW, $jobDescriptor->JobStatus);
+        // Check that our message was added
+        $this->assertStringContainsString('Attempt 1 failed. 3 attempts remaining', $jobDescriptor->SavedJobMessages);
+
+        $jobData = unserialize($jobDescriptor->SavedJobData);
+        $attemptCount = $jobData->AttemptCount ?? 0;
+
+        $this->assertEquals(1, $attemptCount);
+        $this->assertNotNull($jobDescriptor->StartAfter);
+
+        $startAfter = strtotime($jobDescriptor->StartAfter);
+
+        // After the 1st try we except a 2-minutes (120-second) delay for the StartAfter. Putting in a window of 10
+        // seconds for the comparison to reduce flakiness
+        $expectedLow = DBDatetime::now()->getTimestamp() + 115;
```

Comment:
would setting a mock 'now' help this?

Reply:
I still can't guarantee that the process completes within that same second though, so I think a small buffer is going to be required.

EG:

- I set mock now to `2020-01-01 01:00:00`
- So, I'll be expected a new start time of 2 minutes after this for the first failure (`2020-01-01 01:02:00`)
- The job starts processing
- The job takes > 1s to fail
- The job sets the new start time as `2020-01-01 01:02:01`
- Test fails

I could move things around a bit so that the buffer is smaller, but I'm not sure it can be completely avoided, since I can't know the exact second when the job process fails and sets the new start time.

Reply:
ah I see yeah you could end up with fairly useless test failures just as you add logic etc - all good